File: ios/Base/Callback.swift
//
//  Callback.swift
//  ar_rtc_engine
//
//  Created by 余生丶 on 2020/10/29.
//

import Foundation
import ARtcKit

@objc
protocol Callback: class {
    func success(_ data: Any?)

    func failure(_ code: String, _ message: String)
}

extension Callback {
    func code(_ code: Int32?, _ runnable: ((Int32) -> Any?)? = nil) {
        if code == nil || code! < 0 {
            let newCode = abs(Int(code ?? Int32(ARErrorCode.notInitialized.rawValue)))
            failure(String(newCode), ARtcEngineKit.getErrorDescription(newCode) ?? "")
            return
        }

        let res = runnable?(code!)
        if res is Void {
            success(nil)
        } else {
            success(res)
        }
    }

    func resolve<T>(_ source: T?, _ runnable: (T) -> Any?) {
        guard let `source` = source else {
            let code = ARErrorCode.notInitialized.rawValue
            failure(String(code), ARtcEngineKit.getErrorDescription(code) ?? "")
            return
        }

        let res = runnable(source)
        if res is Void {
            success(nil)
        } else {
            success(res)
        }
    }
}

File: android/src/main/java/org/ar/rtc/base/Callback.kt
package org.ar.rtc.base

import org.ar.rtc.Constants
import org.ar.rtc.RtcEngine
import kotlin.math.abs

abstract class Callback {
    fun code(code: Int?, runnable: ((Int) -> Any?)? = null) {
        if (code == null || code < 0) {
            val newCode = abs(code ?: Constants.ERR_NOT_INITIALIZED)
            failure(newCode.toString(), "msg")
            return
        }

        val res = if (runnable != null) runnable(code) else Unit
        if (res is Unit) {
            success(null)
        } else {
            success(res)
        }
    }

    fun <T> resolve(source: T?, runnable: (T) -> Any?) {
        if (source == null) {
            val code = Constants.ERR_NOT_INITIALIZED
            failure(code.toString(), "msg")
            return
        }

        val res = runnable(source)
        if (res is Unit) {
            success(null)
        } else {
            success(res)
        }
    }

    abstract fun success(data: Any?)

    abstract fun failure(code: String, message: String)
}
